### PR TITLE
[eslint] Don't crash on multiple `@babel/parser` copies

### PR DESCRIPTION
**Pull request by nicolo-ribaudo**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes https://github.com/babel/babel/issues/12985, fixes https://github.com/babel/babel/issues/12753
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

This PR replaces object ref comparisons with string comparisons.

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/13274"></a>

### CODE REVIEW

File: eslint/babel-eslint-tests/test/integration/duplicated-babel-parser.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+  it("works with different copies of @babel/parser", () => {
+    const engine = new eslint.CLIEngine({ ignore: false });
+    expect(() =>
+      engine.executeOnFiles([
+        path.resolve(
+          path.dirname(fileURLToPath(import.meta.url)),
+          `../fixtures/duplicated-babel-parser/a.js`,
+        ),
+      ]),
+    ).not.toThrow();
```
Comment (nicolo-ribaudo) [inline eslint/babel-eslint-tests/test/integration/duplicated-babel-parser.js:15]: This throws without the fix (it's the example from https://github.com/babel/babel/issues/12985#issuecomment-795666007)

**Review by JLHwung — COMMENTED** [review]
Can we check whether we have different copies of `@babel/parser` when loading `covertTokens`? Then we can use the slower string match only when we know a fast token match check is not feasible.

**Comment by nicolo-ribaudo** [comment]
When the objects would be equal, the strings are the same (I don't mean that they are `===`, I mean that they are exactly the same string created in the same place at the same time).

v8 internally compares them by pointer, so it's as fast as comparing the objects:
1. The `===` implementation: https://github.com/v8/v8/blob/dc712da548c7fb433caed56af9a021d964952728/src/objects/objects.cc#L816
2. The equality implementation for strings: https://github.com/v8/v8/blob/dc712da548c7fb433caed56af9a021d964952728/src/objects/string-inl.h#L373

**Comment by JLHwung** [comment]
Did you mean https://github.com/v8/v8/blob/dc712da548c7fb433caed56af9a021d964952728/src/objects/objects.cc#L821 ? The token is a plain object while `token.label` is a string, so comparing token is faster than comparing strings because the former essentially compare the memory address.

I agree that they are probably neglectable compared to parse/estraverse. So I am good on this PR.

**Review by JLHwung — APPROVED** [review]

File: eslint/babel-eslint-parser/src/convert/convertTokens.js [hunk 2]
```diff
@@ -1,4 +1,10 @@
-import { tokTypes as tt } from "@babel/core";
+import { tokTypes } from "@babel/core";
+
+const tl = (process.env.BABEL_8_BREAKING
```
Comment (kaicataldo) [inline eslint/babel-eslint-parser/src/convert/convertTokens.js:3]: Do we want to try to figure out a way to test this? Would it work to just manually set `process.env.BABEL_8_BREAKING` in our tests?

**Review by kaicataldo — APPROVED** [review]
One question about increasing test coverage but otherwise LGTM!

File: eslint/babel-eslint-parser/src/convert/convertTokens.js (hunk 2 above)
Comment (nicolo-ribaudo) [reply eslint/babel-eslint-parser/src/convert/convertTokens.js:3]: Yeah, it's tested by the "Test Babel 8 breaking changes" job (even if it doesn't show up in codecov)

**Comment by nicolo-ribaudo** [comment]
> Did you mean v8/v8@dc712da/src/objects/objects.cc#L821 ? The token is a plain object while token.label is a string, so comparing token is faster than comparing strings because the former essentially compare the memory address.

Reading the code, I think that when comparing interned strings it also compares just the memory address (the second of my links), so it should be equally fast (it does three checks: pointer + first is interned + second is interned, but it still doesn't need to iterate the string).

**Pull request merged**